src/ray/raylet/node_manager.cc

```diff
@@ -1405,7 +1405,9 @@ void NodeManager::ProcessFetchOrReconstructMessage(
     if (!worker) {
       worker = worker_pool_.GetRegisteredDriver(client);
     }
-    if (worker) {
+    // Fetch requests can get re-ordered after the worker finishes, so make sure to
+    // check the worker is still assigned a task to avoid leaks.
+    if (worker && !worker->GetAssignedTaskId().IsNil()) {
```

Comment:
This is the most straightforward fix I can think of (checking assigned task id is used to suppress re-ordering anomalies elsewhere already).

Other alternatives including making FetchOrReconstruct a synchronous call, though that could hurt performance. Or, combining FetchAndReconstruct with the plasma Get call.